File: Go/pkg/observe/statistics.go
package observe

/*
	Package 'observe' provides mechanism's for observing and logging system statistics during application runtime.
*/

import (
	"log/slog"
	"strconv"
	"time"
)

/*
	The Statistics module provides methods to track and log timing statistics during application execution.
*/

type Statistics struct {
	startTime time.Time
	endTime   time.Time
}

func NewStatistics() *Statistics {
	slog.SetLogLoggerLevel(slog.LevelInfo)
	return &Statistics{startTime: time.Time{}, endTime: time.Time{}}
}

// StartTime records the start time for statistics measurement.
func (s *Statistics) StartTime() {
	s.startTime = time.Now()
	slog.Debug("Statistics: Start time recorded")
}

// StopTime records the end time for statistics measurement.
func (s *Statistics) StopTime() {
	s.endTime = time.Now()
	slog.Debug("Statistics: End time recorded")
}

// getElapsedTimeSeconds calculates the elapsed time in seconds between start and end times.
func (s *Statistics) getElapsedTimeSeconds() float64 {
	if s.startTime.IsZero() || s.endTime.IsZero() {
		slog.Error("Statistics: Timer has not been started and/or stopped properly")
		return 0
	}
	return s.endTime.Sub(s.startTime).Seconds()
}

// Log outputs the statistics information.
func (s *Statistics) Log() {
	slog.Info("Statistics: Elapsed time=" + strconv.FormatFloat(s.getElapsedTimeSeconds(), 'f', 20, 64) + " seconds")
}

// Reset clears the recorded start and end times.
func (s *Statistics) Reset() {
	s.startTime = time.Time{}
	s.endTime = time.Time{}
	slog.Debug("Statistics: Statistics reset")
}

File: Python/pkg/observe/statistics.py
import time

"""
    Package 'observe' provides mechanism's for observing and logging system statistics during application runtime.
"""

"""
    Objects of class Statistics provides methods to collect and log timing statistics during application execution.
"""

class Statistics:
    def __init__(self, logger):
        # Setup logger
        self.logger = logger
        # Initialize statistics fields
        self.begin_time = None
        self.end_time = None
       

    def start_time(self):
        """
        Records the start time for statistics measurement.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.begin_time = time.time()
        self.logger.debug("Statistics: Start time recorded")

    def stop_time(self):
        """
        Records the end time for statistics measurement.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.end_time = time.time()
        self.logger.debug("Statistics: End time recorded")
        

    def get_elapsed_time(self):
        """
        Calculates the elapsed time in seconds between start and end times.

        Args:
            None
        Returns:
            float: Elapsed time in seconds.
        Raises:
            ValueError: If the timer has not been started and stopped properly.
        """
        if self.begin_time is None or self.end_time is None:
            raise ValueError("Statistics: Timer has not been started and/or stopped properly")
        return (self.end_time - self.begin_time)

    def log(self):  
        """
        Outputs the statistic information.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.logger.info(f"Statistics: Elapsed time={self.get_elapsed_time()} seconds")

    def reset(self):
        """
        Clears the recorded start and end times.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.begin_time = None
        self.end_time = None
        self.logger.debug("Statistics: Statistics reset")